doc/source/reference/array_api.rst

```diff
@@ -0,0 +1,795 @@
+   * - ``matrix_rank``
+     - ``rtol``
+     - ``tol``
+     - The definitions of ``rtol`` and ``tol`` differ. See
+       :ref:`array_api-linear-algebra-differences`.
+   * - ``pinv``
+     - ``rtol``
+     - ``rcond``
+     - The definitions of ``rtol`` and ``rcond`` are the same, but their
+       default values differ. See :ref:`array_api-linear-algebra-differences`.
+   * - ``std`` and ``var``
+     - ``correction``
+     - ``ddof``
+     -
+
+
+.. _array_api-type-promotion-differences:
+
+Type Promotion Differences
+--------------------------
+
+Type promotion is the biggest area where NumPy deviates from the spec. The
+most notable difference is that NumPy does value-based casting in many cases.
+The spec explicitly disallows value-based casting. In the array API, the
+result type of any operation is always determined entirely by the input types,
+independently of values or shapes.
+
+.. list-table::
+   :header-rows: 1
+
+   * - Feature
+     - Type
+     - Notes
+   * - Limited set of dtypes.
+     - **Strictness**
+     - ``numpy.array_api`` only implements those `dtypes that are required by
+       the spec
+       <https://data-apis.org/array-api/latest/API_specification/data_types.html>`__.
+   * - Operators (like ``+``) with Python scalars only accept matching
+       scalar types.
+     - **Strictness**
+     - For example, ``<int32 array> + 1.0`` is not allowed. See `the spec
+       rules for mixing arrays and Python scalars
+       <https://data-apis.org/array-api/latest/API_specification/type_promotion.html#mixing-arrays-with-python-scalars>`__.
+   * - Operators (like ``+``) with Python scalars always return the same dtype
+       as the array.
+     - **Breaking**
+     - For example, ``numpy.array_api.asarray(0., dtype=float32) + 1e64`` is a
+       ``float32`` array.
+   * - In-place operators are disallowed when the left-hand side would be
+       promoted.
+     - **Breaking**
+     - Example: ``a = np.array(1, dtype=np.int8); a += np.array(1, dtype=np.int16)``. The spec explicitly disallows this.
+   * - ``int`` promotion for operators is only specified for integers within
+       the bounds of the dtype.
+     - **Strictness**
+     - ``numpy.array_api`` fallsback to ``np.ndarray`` behavior (either
+       cast or raise ``OverflowError``).
+   * - ``__pow__`` and ``__rpow__`` do not do value-based casting for 0-D
+       arrays.
+     - **Breaking**
+     - For example, ``np.array(0., dtype=float32)*np.array(0.,
+       dtype=float64)`` is ``float32``. Note that this is value-based casting
+       on 0-D arrays, not scalars.
+   * - No cross-kind casting.
+     - **Strictness**
+     - Namely, boolean, integer, and floating-point data types do not cast to
+       each other, except explicitly with ``astype`` (this is separate from
+       the behavior with Python scalars).
+   * - No casting unsigned integer dtypes to floating dtypes (e.g., ``int64 +
+       uint64 -> float64``.
+     - **Strictness**
```

Comment:
Yeah, I was also unsure about this, but other folks here seemed to agree that the spec does not disallow any additional type promotions, including this one (CC @rgommers @kgryte).

Reply:
I think it's not breaking. There is no `uint64 + int64` promotion rule, which means it is undefined behavior. Libraries may decide to raise an exception, cast to `float64`, cast to `int64`, or do whatever else they like.

Reply:
>  I guess that distinction would only mainly matter if there is a chance that you may want to specify contradicting promotions it in the future.

That's indeed the one/main reason to specify particular exceptions or "implementing this is not allowed". We've tried to stay away from specifying exceptions or error handling type things, because it's impossible to be comprehensive (in addition for it to be difficult to align libraries' behavior here).